src/platform/packages/shared/kbn-apm-synthtrace/src/cli/utils/stream_manager.ts

```diff
@@ -98,21 +99,34 @@ export class StreamManager {
     generator: SynthGenerator<Fields>
   ): Promise<void> {
     const clientStream = this.createOrReuseClientStream(client);
+    const generatorStream = new PassThrough({ objectMode: true });
+    const generatorArray = castArray(generator);
+
+    generatorArray.reverse().forEach((current) => {
+      const currentStream = isGeneratorObject(current) ? Readable.from(current) : current;
+      currentStream.pipe(generatorStream, { end: false });
+      this.trackedStreams.push(currentStream);
+
+      currentStream.on('end', () => {
+        pull(this.trackedStreams, currentStream);
+        if (this.trackedStreams.length === 0) {
+          generatorStream.end();
+        }
+      });
 
-    const generatorStream = castArray(generator)
-      .reverse()
-      .reduce<Writable>((prev, current) => {
-        const currentStream = isGeneratorObject(current) ? Readable.from(current) : current;
-        return currentStream.pipe(prev);
-      }, new PassThrough({ objectMode: true }));
+      currentStream.on('error', (err) => {
+        pull(this.trackedStreams, currentStream);
+        generatorStream.destroy(err);
+      });
+    });
 
```

Comment:
`currentStream.pipe(prev)` with `end: true` was causing the previous piped stream to be prematurely closed

Reply:
isn't that the only change we need to make then?

Reply:
we still need to end the `generatorStream` otherwise it will be opened forever